File: decision_tree_regression.R
# Decision Tree Regression

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# Not splitting here due to small dataset size (10 samples).
# In production, use:
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 2/3)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# Decision Trees do not require feature scaling.

# Fitting Decision Tree Regression to the dataset
library(rpart)
regressor = rpart(formula = Salary ~ .,
                  data = dataset,
                  control = rpart.control(minsplit = 1))

# Predicting a new result with Decision Tree Regression
y_pred = predict(regressor, data.frame(Level = 6.5))
cat(sprintf("Predicted salary for level 6.5: %s\n", format(y_pred, big.mark = ",")))

# Visualising the Decision Tree Regression results (higher resolution)
library(ggplot2)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.01)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(regressor, newdata = data.frame(Level = x_grid))),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Decision Tree Regression)') +
  xlab('Level') +
  ylab('Salary')

# Plotting the tree
plot(regressor)
text(regressor)

File: decision_tree_regression.py
# Decision Tree Regression

# Importing the libraries
import os
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.tree import DecisionTreeRegressor


def main():
    # Importing the dataset (resolve path relative to this script)
    script_dir = os.path.dirname(os.path.abspath(__file__))
    csv_path = os.path.join(script_dir, "Position_Salaries.csv")
    dataset = pd.read_csv(csv_path)
    X = dataset.iloc[:, 1:2].values
    y = dataset.iloc[:, 2].values

    # Fitting Decision Tree Regression to the dataset
    regressor = DecisionTreeRegressor(random_state=0)
    regressor.fit(X, y)

    # Predicting a new result (predict expects a 2D array)
    y_pred = regressor.predict([[6.5]])
    print(f"Predicted salary for level 6.5: {y_pred[0]:,.0f}")

    # Visualising the Decision Tree Regression results (higher resolution)
    X_grid = np.arange(min(X), max(X), 0.01)
    X_grid = X_grid.reshape((len(X_grid), 1))
    plt.scatter(X, y, color="red")
    plt.plot(X_grid, regressor.predict(X_grid), color="blue")
    plt.title("Truth or Bluff (Decision Tree Regression)")
    plt.xlabel("Position level")
    plt.ylabel("Salary")
    plt.show()


if __name__ == "__main__":
    main()
